# 🧩 Fase 1 - Initial Data Exploration Script (EDA) and Data Cleaning


### 1. **Library imports**  


In [57]:
import pandas as pd
import numpy as np

### 2. **Dataset loading and configuration**  


In [58]:
# Dataframe: Customer Flight Activity
df_flight_original = pd.read_csv("Customer Flight Activity.csv")
df_flight = df_flight_original.copy() # Creates a backup before modifications
# Dataframe: Customer Loyalt History
df_loyalty = pd.read_csv("Customer Loyalty History.csv")
# Configuration
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

### 3. **Basic structural overview** (`shape`, `size`,`info`, `head`, `tail`)  


**Dataframe: Customer Flight Activity**

In [59]:
df_flight.shape

(405624, 10)

In [60]:
df_flight.columns

Index(['Loyalty Number', 'Year', 'Month', 'Flights Booked',
       'Flights with Companions', 'Total Flights', 'Distance',
       'Points Accumulated', 'Points Redeemed', 'Dollar Cost Points Redeemed'],
      dtype='object')

In [61]:
df_flight.head()


Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
0          100018  2017      1               3                        0   
1          100102  2017      1              10                        4   
2          100140  2017      1               6                        0   
3          100214  2017      1               0                        0   
4          100272  2017      1               0                        0   

   Total Flights  Distance  Points Accumulated  Points Redeemed  \
0              3      1521               152.0                0   
1             14      2030               203.0                0   
2              6      1200               120.0                0   
3              0         0                 0.0                0   
4              0         0                 0.0                0   

   Dollar Cost Points Redeemed  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0

In [62]:
df_flight.tail()

Loyalty Number  Year  Month  Flights Booked  Flights with Companions  \
405619          999902  2018     12               0                        0   
405620          999911  2018     12               0                        0   
405621          999940  2018     12               3                        0   
405622          999982  2018     12               0                        0   
405623          999986  2018     12               0                        0   

        Total Flights  Distance  Points Accumulated  Points Redeemed  \
405619              0         0                 0.0                0   
405620              0         0                 0.0                0   
405621              3      1233               123.0                0   
405622              0         0                 0.0                0   
405623              0         0                 0.0                0   

        Dollar Cost Points Redeemed  
405619                            0  
405620                            0  
405621                            0  
405622                            0  
405623                            0

In [63]:
df_flight.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


**Dataframe: Customer Loyalt History**

In [64]:
df_loyalty.shape

(16737, 16)

In [65]:
df_loyalty.head()


Loyalty Number Country          Province       City Postal Code  Gender  \
0          480934  Canada           Ontario    Toronto     M2Z 4K1  Female   
1          549612  Canada           Alberta   Edmonton     T3G 6Y6    Male   
2          429460  Canada  British Columbia  Vancouver     V6E 3D9    Male   
3          608370  Canada           Ontario    Toronto     P1W 1K4    Male   
4          530508  Canada            Quebec       Hull     J8Y 3Z5    Male   

  Education    Salary Marital Status Loyalty Card      CLV Enrollment Type  \
0  Bachelor   83236.0        Married         Star  3839.14        Standard   
1   College       NaN       Divorced         Star  3839.61        Standard   
2   College       NaN         Single         Star  3839.75        Standard   
3   College       NaN         Single         Star  3839.75        Standard   
4  Bachelor  103495.0        Married         Star  3842.79        Standard   

   Enrollment Year  Enrollment Month  Cancellation Year  Cancellation Month  
0             2016                 2                NaN                 NaN  
1             2016                 3                NaN                 NaN  
2             2014                 7             2018.0                 1.0  
3             2013                 2                NaN                 NaN  
4             2014                10                NaN                 NaN

In [66]:
df_loyalty.tail()

Loyalty Number Country          Province        City Postal Code  \
16732          823768  Canada  British Columbia   Vancouver     V6E 3Z3   
16733          680886  Canada      Saskatchewan      Regina     S1J 3C5   
16734          776187  Canada  British Columbia   Vancouver     V5R 1W3   
16735          906428  Canada             Yukon  Whitehorse     Y2K 6R0   
16736          652627  Canada          Manitoba    Winnipeg     R2C 0M5   

       Gender Education   Salary Marital Status Loyalty Card       CLV  \
16732  Female   College      NaN        Married         Star  61850.19   
16733  Female  Bachelor  89210.0        Married         Star  67907.27   
16734    Male   College      NaN         Single         Star  74228.52   
16735    Male  Bachelor -57297.0        Married         Star  10018.66   
16736  Female  Bachelor  75049.0        Married         Star  83325.38   

      Enrollment Type  Enrollment Year  Enrollment Month  Cancellation Year  \
16732        Standard             2012                12                NaN   
16733        Standard             2014                 9                NaN   
16734        Standard             2014                 3                NaN   
16735  2018 Promotion             2018                 4                NaN   
16736        Standard             2015                12             2016.0   

       Cancellation Month  
16732                 NaN  
16733                 NaN  
16734                 NaN  
16735                 NaN  
16736                 8.0

In [67]:
df_loyalty.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  object 
 2   Province            16737 non-null  object 
 3   City                16737 non-null  object 
 4   Postal Code         16737 non-null  object 
 5   Gender              16737 non-null  object 
 6   Education           16737 non-null  object 
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  object 
 9   Loyalty Card        16737 non-null  object 
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  object 
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: f

### 4. **Standardization of column names** 


In [68]:
df_flight.columns = df_flight.columns.str.strip().str.lower()
df_loyalty.columns = df_loyalty.columns.str.strip().str.lower()

df_loyalty.head(1)

loyalty number country province     city postal code  gender education  \
0          480934  Canada  Ontario  Toronto     M2Z 4K1  Female  Bachelor   

    salary marital status loyalty card      clv enrollment type  \
0  83236.0        Married         Star  3839.14        Standard   

   enrollment year  enrollment month  cancellation year  cancellation month  
0             2016                 2                NaN                 NaN

### 5. **Detection of missing values** 


In [69]:
# Flight DataFrame -> NO null values
print(df_flight.isnull().sum())
print('--------------------------')
print((df_flight.isnull().mean() * 100).round(2))


loyalty number                 0
year                           0
month                          0
flights booked                 0
flights with companions        0
total flights                  0
distance                       0
points accumulated             0
points redeemed                0
dollar cost points redeemed    0
dtype: int64
--------------------------
loyalty number                 0.0
year                           0.0
month                          0.0
flights booked                 0.0
flights with companions        0.0
total flights                  0.0
distance                       0.0
points accumulated             0.0
points redeemed                0.0
dollar cost points redeemed    0.0
dtype: float64


In [70]:
# Loyalty DataFrame
print(df_loyalty.isnull().sum().sort_values(ascending=False))
print('--------------------------')
print((df_loyalty.isnull().mean() * 100).sort_values(ascending=False).round(2))
print("cancellation year and cancellation month have a large ammount of nulls")
print("salary has 25% of nulls")

cancellation month    14670
cancellation year     14670
salary                 4238
loyalty number            0
city                      0
postal code               0
province                  0
country                   0
education                 0
gender                    0
marital status            0
loyalty card              0
enrollment type           0
clv                       0
enrollment month          0
enrollment year           0
dtype: int64
--------------------------
cancellation month    87.65
cancellation year     87.65
salary                25.32
loyalty number         0.00
city                   0.00
postal code            0.00
province               0.00
country                0.00
education              0.00
gender                 0.00
marital status         0.00
loyalty card           0.00
enrollment type        0.00
clv                    0.00
enrollment month       0.00
enrollment year        0.00
dtype: float64
cancellation year and cancellation month have a l

### **Cleaning of missing values** 


Los valores nulos en las columnas `cancellation month` y `cancellation year` probablemente indican que el cliente **no ha cancelado** su participación en el programa de fidelidad. 

Por lo tanto, no se trata de un error en los datos, sino de una información implícita sobre la vigencia del cliente.

Esto podría ser la razón del alto porcentaje de valores nulos en ambas columnas:


- `cancellation month`: 87.65% de los registros
- `cancellation year`: 87.65% de los registros


In [71]:
# Create a new boolean column indicating if the customer cancelled
df_loyalty['cancelled'] = df_loyalty['cancellation year'].notna()
df_loyalty

# Returns True where 'cancellation year' is not null.
# Returns False where it is null.

# .notna() was used because the goal was to identify customers who cancelled — 
# and the presence of a value in the 'cancellation year' column indicates that cancellation -> (True).

# .isnull() would do the opposite: it would return True for active (non-cancelled) customers, 
# which can also be useful in other contexts.


loyalty number country          province        city postal code  \
0              480934  Canada           Ontario     Toronto     M2Z 4K1   
1              549612  Canada           Alberta    Edmonton     T3G 6Y6   
2              429460  Canada  British Columbia   Vancouver     V6E 3D9   
3              608370  Canada           Ontario     Toronto     P1W 1K4   
4              530508  Canada            Quebec        Hull     J8Y 3Z5   
...               ...     ...               ...         ...         ...   
16732          823768  Canada  British Columbia   Vancouver     V6E 3Z3   
16733          680886  Canada      Saskatchewan      Regina     S1J 3C5   
16734          776187  Canada  British Columbia   Vancouver     V5R 1W3   
16735          906428  Canada             Yukon  Whitehorse     Y2K 6R0   
16736          652627  Canada          Manitoba    Winnipeg     R2C 0M5   

       gender education    salary marital status loyalty card       clv  \
0      Female  Bachelor   83236.0        Married         Star   3839.14   
1        Male   College       NaN       Divorced         Star   3839.61   
2        Male   College       NaN         Single         Star   3839.75   
3        Male   College       NaN         Single         Star   3839.75   
4        Male  Bachelor  103495.0        Married         Star   3842.79   
...       ...       ...       ...            ...          ...       ...   
16732  Female   College       NaN        Married         Star  61850.19   
16733  Female  Bachelor   89210.0        Married         Star  67907.27   
16734    Male   College       NaN         Single         Star  74228.52   
16735    Male  Bachelor  -57297.0        Married         Star  10018.66   
16736  Female  Bachelor   75049.0        Married         Star  83325.38   

      enrollment type  enrollment year  enrollment month  cancellation year  \
0            Standard             2016                 2                NaN   
1            Standard             2016                 3                NaN   
2            Standard             2014                 7             2018.0   
3            Standard             2013                 2                NaN   
4            Standard             2014                10                NaN   
...               ...              ...               ...                ...   
16732        Standard             2012                12                NaN   
16733        Standard             2014                 9                NaN   
16734        Standard             2014                 3                NaN   
16735  2018 Promotion             2018                 4                NaN   
16736        Standard             2015                12             2016.0   

       cancellation month  cancelled  
0                     NaN      False  
1                     NaN      False  
2                     1.0       True  
3                     NaN      False  
4                     NaN      False  
...                   ...        ...  
16732                 NaN      False  
16733                 NaN      False  
16734                 NaN      False  
16735                 NaN      False  
16736                 8.0       True  

[16737 rows x 17 columns]

Tomé la decisión de eliminar las columnas `cancellation year` y `cancellation month` porque:

- Ya hemos extraído la información esencial mediante la variable booleana `cancelled`, que indica si el cliente ha cancelado o no.
- Ambas columnas presentan un alto porcentaje de valores nulos y ya no aportan valor directo para los análisis o visualizaciones.
- Mantener el DataFrame más limpio y conciso facilita su manipulación, interpretación y visualización.

Además, en caso de que el enfoque del análisis se centre exclusivamente en los clientes vigentes (no cancelados), también sería válido eliminar los registros donde `cancelled` es `True`.


In [72]:
df_loyalty = df_loyalty.drop(columns=['cancellation year', 'cancellation month'])  # elimina as duas colunas permanentemente

# .drop(): method used to remove rows or columns
# columns=[...]: specifies the list of column names to remove
# axis=1 is implicit when using columns=, so no need to write it here
# the result is assigned back to df_loyalty to overwrite the original dataframe

In [73]:
df_loyalty.head(1)

loyalty number country province     city postal code  gender education  \
0          480934  Canada  Ontario  Toronto     M2Z 4K1  Female  Bachelor   

    salary marital status loyalty card      clv enrollment type  \
0  83236.0        Married         Star  3839.14        Standard   

   enrollment year  enrollment month  cancelled  
0             2016                 2      False

Opciones de imputación para la variable `salary`

- **Media**: apropiada si la distribución del salario es simétrica y no presenta valores atípicos extremos.

- **Mediana**: recomendable cuando existen valores atípicos, ya que no se ve afectada por ellos. 

- **Imputación por grupo**: consiste en calcular el promedio (o la mediana) del salario dentro de subgrupos, como por ejemplo:
  - Rango de edad (si existiera la variable)
  - Nivel educativo
  - Nivel de fidelidad o tipo de tarjeta

Esto permite una imputación más contextualizada y representativa para cada segmento de clientes.


In [74]:
# Exploring unique values in 'salary' to guide our decision-making
df_loyalty['salary'].unique()

array([ 83236.,     nan, 103495., ...,  76178.,  91970., -57297.],
      shape=(5891,))

In [75]:
# Show all unique values in 'salary', sorted in ascending order
# Negative values are detected
sorted(df_loyalty['salary'].unique())


[np.float64(-58486.0),
 np.float64(-49830.0),
 np.float64(-46683.0),
 np.float64(-45962.0),
 np.float64(-43234.0),
 np.float64(-19325.0),
 np.float64(15609.0),
 np.float64(16853.0),
 np.float64(19464.0),
 np.float64(21616.0),
 np.float64(23136.0),
 np.float64(26880.0),
 np.float64(26962.0),
 np.float64(28464.0),
 np.float64(33651.0),
 np.float64(33784.0),
 np.float64(34210.0),
 np.float64(35424.0),
 np.float64(39025.0),
 np.float64(39079.0),
 np.float64(39943.0),
 np.float64(40070.0),
 np.float64(40125.0),
 np.float64(40157.0),
 np.float64(41120.0),
 np.float64(41957.0),
 np.float64(42867.0),
 np.float64(42914.0),
 np.float64(43280.0),
 np.float64(43350.0),
 np.float64(44112.0),
 np.float64(44595.0),
 np.float64(45580.0),
 np.float64(45654.0),
 np.float64(46444.0),
 np.float64(46548.0),
 np.float64(46580.0),
 np.float64(46604.0),
 np.float64(46607.0),
 np.float64(46663.0),
 np.float64(46699.0),
 np.float64(46768.0),
 np.float64(46811.0),
 np.float64(46813.0),
 np.float64(46863.0),
 np.

### 🔴 Valor de Salario Negativo ###


| Possibles Causas                            | Explicación                                                                                          |
| ------------------------------------------- | ---------------------------------------------------------------------------------------------------- |
| **Error de tipeo**                          | Alguien ingresó el valor con signo `-` por error.                                                    |
| **Codificación incorrecta**                 | Algunos sistemas usan valores negativos para marcar "desconocido", "no informado", etc.              |
| **Error en conversión de moneda o cálculo** | Errores en fórmulas de conversión o descuentos aplicados incorrectamente durante el procesamiento.   |
| **Problemas de limpieza previa**            | Datos corruptos al combinar múltiples bases o al realizar transformaciones sin tratamiento adecuado. |


In [76]:
# This filters the dataframe to include only rows where the salary is less than 0 (negative values)
df_loyalty[df_loyalty['salary'] < 0]  

# A pattern was observed where most of the negative salary values are associated with a similar education level. 
# However, their salary amounts vary significantly, ranging from 9,081 to 58,486. 
# Therefore, a group-based imputation method will not be applied.
# Additionally, there are repeated values in the same month and year, which may suggest a potential data entry or imputation error.


loyalty number country          province            city postal code  \
1082           542976  Canada            Quebec        Montreal     H2Y 4R4   
1894           959977  Canada  British Columbia       Vancouver     V5R 1W3   
2471           232755  Canada  British Columbia       Vancouver     V1E 4R6   
3575           525245  Canada  British Columbia        Victoria     V10 6T5   
3932           603070  Canada  British Columbia  West Vancouver     V6V 8Z3   
4712           491242  Canada  British Columbia    Dawson Creek     U5I 4F1   
6560           115505  Canada      Newfoundland      St. John's     A1C 6H9   
6570           430398  Canada      Newfoundland      St. John's     A1C 6H9   
7373           152016  Canada           Ontario         Toronto     P1J 8T7   
8576           194065  Canada           Ontario         Sudbury     M5V 1G5   
8767           212128  Canada           Ontario         Toronto     P2T 6G3   
10232          790475  Canada           Ontario         Trenton     K8V 4B2   
11635          366599  Canada           Ontario         Toronto     M1R 4K3   
12596          436087  Canada            Quebec        Montreal     H2T 9K8   
13564          364596  Canada            Quebec       Tremblant     H5Y 2S9   
14327          239955  Canada            Quebec            Hull     J8Y 3Z5   
14355          347013  Canada            Quebec     Quebec City     G1B 3L5   
15416          729561  Canada            Quebec     Quebec City     G1B 3L5   
16431          734647  Canada      Saskatchewan          Regina     S1J 3C5   
16735          906428  Canada             Yukon      Whitehorse     Y2K 6R0   

       gender             education   salary marital status loyalty card  \
1082     Male  High School or Below -49830.0       Divorced         Star   
1894   Female              Bachelor -12497.0        Married       Aurora   
2471   Female              Bachelor -46683.0         Single         Nova   
3575     Male              Bachelor -45962.0        Married         Star   
3932   Female              Bachelor -19325.0         Single         Star   
4712     Male              Bachelor -43234.0        Married         Star   
6560     Male              Bachelor -10605.0        Married         Nova   
6570     Male              Bachelor -17534.0        Married         Nova   
7373   Female              Bachelor -58486.0        Married       Aurora   
8576   Female              Bachelor -31911.0        Married         Nova   
8767     Male              Bachelor -49001.0        Married         Nova   
10232  Female              Bachelor -34079.0        Married         Nova   
11635  Female              Bachelor  -9081.0        Married         Star   
12596    Male              Bachelor -46470.0        Married         Star   
13564  Female              Bachelor -26322.0         Single       Aurora   
14327  Female              Bachelor -47310.0        Married         Nova   
14355    Male              Bachelor -39503.0        Married         Nova   
15416  Female              Bachelor -19332.0       Divorced         Star   
16431    Male              Bachelor -46303.0        Married         Nova   
16735    Male              Bachelor -57297.0        Married         Star   

            clv enrollment type  enrollment year  enrollment month  cancelled  
1082   24127.50  2018 Promotion             2018                 3      False  
1894    9453.00  2018 Promotion             2018                 3      False  
2471    4787.81  2018 Promotion             2018                 3      False  
3575    2402.33  2018 Promotion             2018                 3      False  
3932    2893.74  2018 Promotion             2018                 3      False  
4712    7597.91  2018 Promotion             2018                 3      False  
6560    5860.17  2018 Promotion             2018                 4      False  
6570   49423.80  2018 Promotion             2018                 3      False  
7373    5067.21  2018 Promotio

### Tratamiento de Salarios Negativos

Como se han identificado **20 salarios negativos**, tomamos la decisión de **reemplazarlos por valores nulos (`NaN`)**, ya que no tienen sentido lógico en este contexto.

Posteriormente, todos los valores `NaN` en la columna `salary` serán imputados utilizando una estrategia adecuada. 

In [77]:
# Get the unique values among all negative salaries
df_loyalty[df_loyalty['salary'] < 0]['salary'].unique()

# Count how many rows have negative salary values: 20 rowns
df_loyalty[df_loyalty['salary'] < 0].shape[0]
# .shape returns (100, 10)
# .shape[0]Returns only the number of rows (first element of the tuple)
# .shape[1]Returns only the number of columns (second element of the tuple)
# Then .shape[0] tells you how many rows match that condition


20

In [78]:
# Analyze salary values before replacing them
print('Average salary (rounded):', round(df_loyalty['salary'].mean(), 2))    
print('Median salary (rounded):', round(df_loyalty['salary'].median(), 2))   

Average salary (rounded): 79245.61
Median salary (rounded): 73455.0


In [79]:
# Replace negative salary values with NaN
df_loyalty.loc[df_loyalty['salary'] < 0, 'salary'] = np.nan

# .loc[] is used to access or modify specific rows and columns in a DataFrame using labels or conditions (df.loc[rows, columns]).
# df_loyalty['salary'] < 0 creates a condition that returns True for rows where the salary is negative [False, True, False, True, ...].
# np.nan comes from the NumPy library and represents a missing value (NaN = Not a Number.
# Final result: negative salary values are replaced by NaN for future imputation.


Después de **reemplazar los valores negativos por valores nulos (`NaN`)**, imputaré la mediana, ya que no se ve afectada por los outliers. 

Esto es importante porque existen sueldos de 15.609 hasta 407.228. En la grafica adjunta podemos ver claramente. 


In [80]:
# Print summary statistics after excluding negative values
print('Average salary (rounded):', round(df_loyalty['salary'].mean(), 2))    
print('Median salary (rounded):', round(df_loyalty['salary'].median(), 2))  
print('Minimum salary:', round(df_loyalty['salary'].min(), 2))     
print('Maximum salary:', round(df_loyalty['salary'].max(), 2))     

Average salary (rounded): 79429.57
Median salary (rounded): 73510.0
Minimum salary: 15609.0
Maximum salary: 407228.0


La media está distorsionada, y la mediana representa mejor el comportamiento central de la mayoría de los clientes.

In [81]:
# Imputar com a mediana 
df_loyalty['salary'] = df_loyalty['salary'].fillna(df_loyalty['salary'].median())

### 6. **Detection of duplicate rows**  


**Dataframe: Customer Flight Activity**


In [82]:
# 1864 fully duplicated rows
df_flight.duplicated().sum()

np.int64(1864)

In [83]:
df_flight[df_flight.duplicated()]  # Show all rows that are 100% identical (full duplicates)
# Since some values can naturally repeat without indicating a problem,
# I decided to investigate whether the same 'loyalty number' appears multiple times.

loyalty number  year  month  flights booked  flights with companions  \
42              101902  2017      1               0                        0   
227             112142  2017      1               0                        0   
478             126100  2017      1               0                        0   
567             130331  2017      1               0                        0   
660             135421  2017      1               0                        0   
...                ...   ...    ...             ...                      ...   
404668          949628  2018     12               0                        0   
404884          960050  2018     12               0                        0   
405111          971370  2018     12               0                        0   
405410          988392  2018     12               0                        0   
405437          989528  2018     12               0                        0   

        total flights  distance  points accumulated  points redeemed  \
42                  0         0                 0.0                0   
227                 0         0                 0.0                0   
478                 0         0                 0.0                0   
567                 0         0                 0.0                0   
660                 0         0                 0.0                0   
...               ...       ...                 ...              ...   
404668              0         0                 0.0                0   
404884              0         0                 0.0                0   
405111              0         0                 0.0                0   
405410              0         0                 0.0                0   
405437              0         0                 0.0                0   

        dollar cost points redeemed  
42                                0  
227                               0  
478                               0  
567                               0  
660                               0  
...                             ...  
404668                            0  
404884                            0  
405111                            0  
405410                            0  
405437                            0  

[1864 rows x 10 columns]

In [84]:
loyalty_duplicate_lines = df_flight[df_flight.duplicated(subset='loyalty number', keep=False)] 
loyalty_duplicate_lines.head()
# Checks for duplicate rows, but only based on the 'loyalty number' column.
# The variation among duplicates seems normal (no immediate issue).

loyalty number  year  month  flights booked  flights with companions  \
0          100018  2017      1               3                        0   
1          100102  2017      1              10                        4   
2          100140  2017      1               6                        0   
3          100214  2017      1               0                        0   
4          100272  2017      1               0                        0   

   total flights  distance  points accumulated  points redeemed  \
0              3      1521               152.0                0   
1             14      2030               203.0                0   
2              6      1200               120.0                0   
3              0         0                 0.0                0   
4              0         0                 0.0                0   

   dollar cost points redeemed  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0

In [85]:
print(df_flight['loyalty number'].value_counts()) # shows how many times each customer appears.
# The methods value_counts() and groupby().size() return the same results when applied to a single column.

# 388,887 duplicated values in the 'loyalty number' column
print('Rows in df_flight where the loyalty number has already appeared before:',df_flight['loyalty number'].duplicated().sum())

loyalty number
678205    72
833793    48
727091    48
700650    48
654433    48
          ..
100590    24
100642    24
100644    24
100646    24
999891    24
Name: count, Length: 16737, dtype: int64
Rows in df_flight where the loyalty number has already appeared before: 388887


### Exploración más detallada para verificar si el mismo número de fidelidad aparece en el mismo mes y año, ya que la duplicación en otras columnas parece esperada.


In [86]:
duplicated_flight = df_flight[df_flight['loyalty number'].duplicated(keep=False)] 
duplicated_flight.sort_values('loyalty number').head()
# keep=False highlights both duplicates and their original occurrences.
# Checks for duplicates within the 'loyalty number' Series.

loyalty number  year  month  flights booked  flights with companions  \
257159          100018  2018      3               7                        0   
257382          100018  2018      2               6                        0   
321119          100018  2018      8               0                        0   
185911          100018  2017     12               6                        0   
270416          100018  2018      5               0                        0   

        total flights  distance  points accumulated  points redeemed  \
257159              7      1876               280.5                0   
257382              6      1824               273.0                0   
321119              0         0                 0.0                0   
185911              6      1908               190.0                0   
270416              0         0                 0.0                0   

        dollar cost points redeemed  
257159                            0  
257382                            0  
321119                            0  
185911                            0  
270416                            0

In [87]:
duplicated_flight[duplicated_flight['loyalty number'] == 100018] 
# I used a sample of customer 100018 from the previous dataframe to explore the data

duplicated_flight[duplicated_flight['loyalty number'] == 100018].sort_values(by=['year', 'month'], ascending=[True, True])
## by=['year', 'month'] → sorts first by year, then by month
# ascending=[True, True] → sorts both in ascending order

loyalty number  year  month  flights booked  flights with companions  \
0               100018  2017      1               3                        0   
16901           100018  2017      2               2                        2   
213311          100018  2017      3              14                        3   
50703           100018  2017      4               4                        0   
67604           100018  2017      5               0                        0   
84505           100018  2017      6               4                        1   
101406          100018  2017      7              10                        0   
142774          100018  2017      8              10                        5   
135208          100018  2017      9               9                        4   
152109          100018  2017     10               6                        0   
169010          100018  2017     11              13                        1   
185911          100018  2017     12               6                        0   
202812          100018  2018      1               1                        0   
257382          100018  2018      2               6                        0   
257159          100018  2018      3               7                        0   
267063          100018  2018      4               5                        0   
270416          100018  2018      5               0                        0   
287317          100018  2018      6               0                        0   
304218          100018  2018      7               9                        7   
321119          100018  2018      8               0                        0   
338020          100018  2018      9              14                        2   
46353           100018  2018     10               6                        4   
371822          100018  2018     11              11                        0   
388723          100018  2018     12              17                        6   

        total flights  distance  points accumulated  points redeemed  \
0                   3      1521               152.0                0   
16901               4      1320               132.0                0   
213311             17      2533               253.0              438   
50703               4       924                92.0                0   
67604               0         0                 0.0                0   
84505               5      4330               433.0                0   
101406             10      3870               387.0                0   
142774             15      4200               420.0              690   
135208             13      1651               165.0                0   
152109              6      1032               103.0                0   
169010             14      3374               337.0                0   
185911              6      1908               190.0                0   
202812              1       312                31.0                0   
257382              6      1824               273.0                0   
257159              7      1876               280.5                0   
267063              5      2590               388.5                0   
270416              0         0                 0.0                0   
287317              0         0                 0.0                0   
304218             16      2736               273.0                0   
321119              0         0                 0.0                0   
338020             16      3568               356.0                0   
46353              10      3110               311.0              385   
371822             11      3058               305.0                0   
388723             23      4945               494.0                0   

        dollar cost points redeemed  
0                                 0  
16901                             0  
213311                           36  
50703                             0  
67604           

El cliente 100018 aparece:

Una vez por mes en los años 2017 y 2018

Con variaciones en 'flights booked', 'distance', 'points accumulated', etc.

Cada fila parece representar actividades mensuales de vuelo

Por lo tanto: no son duplicados exactos, sino registros separados por mes.

Conclusión: estos datos no deben eliminarse como duplicados.

Cada fila representa un vuelo o conjunto de vuelos correspondientes a un mes específico.

El hecho de que el 'loyalty number' se repita solo indica que el cliente voló más de una vez a lo largo del tiempo.

Proximos pasos: verificar si hay repeticion de los mismo loyalt number en los mismos meses y años 

In [88]:
# To explore if there are fully duplicated entries (e.g., same customer, same year and month, and identical data)
print(df_flight.duplicated(subset=['loyalty number', 'year', 'month'], keep=False).sum())
# The value for 'total flights' should not be different for the same customer in the same year and in the same month
duplicated_flight[duplicated_flight.duplicated(subset=['loyalty number', 'year', 'month'], keep=False)]

7848


loyalty number  year  month  flights booked  flights with companions  \
41              101902  2017      1               0                        0   
42              101902  2017      1               0                        0   
114             106001  2017      1               9                        4   
115             106001  2017      1               5                        0   
226             112142  2017      1               0                        0   
...                ...   ...    ...             ...                      ...   
405437          989528  2018     12               0                        0   
405451          990512  2018     12              15                        5   
405452          990512  2018     12               5                        2   
405486          992168  2018     12              15                        0   
405487          992168  2018     12               6                        0   

        total flights  distance  points accumulated  points redeemed  \
41                  0         0                 0.0                0   
42                  0         0                 0.0                0   
114                13      1365               136.0                0   
115                 5       910                91.0                0   
226                 0         0                 0.0                0   
...               ...       ...                 ...              ...   
405437              0         0                 0.0                0   
405451             20      2940               294.0                0   
405452              7      3906               390.0                0   
405486             15      3120               312.0                0   
405487              6      3192               319.0                0   

        dollar cost points redeemed  
41                                0  
42                                0  
114                               0  
115                               0  
226                               0  
...                             ...  
405437                            0  
405451                            0  
405452                            0  
405486                            0  
405487                            0  

[7848 rows x 10 columns]

### 🔴 Valor Duplicado con el mismo Loyalt Number, Mes y Año ###

Detecté registros duplicados que compartían el mismo cliente, año y mes, pero con valores diferentes en columnas como 'total flights'. 

Dado que estas filas se refieren al mismo periodo, no deberían tener discrepancias. 

Esto indica un posible error en la carga o duplicación del sistema. Por lo tanto, decidí agruparlos por cliente, año y mes, consolidando los datos (sumando o tomando el máximo según el campo) para obtener una versión limpia y coherente del dataset.

Probé algunos valores de la columna 'loyalty number' que se repiten, para explorar con más detalle algunos casos específicos:


In [89]:
duplicated_flight[(duplicated_flight['loyalty number'] == 992168) &(duplicated_flight['month'] == 10) &(duplicated_flight['year'] == 2018)]

# Loyalty numbers tested:

# 106001  5 2017
# 101902  1 2017
# 101902  4 2017 
# 992168 10 2018 


loyalty number  year  month  flights booked  flights with companions  \
371684          992168  2018     10               1                        1   
371685          992168  2018     10               0                        0   

        total flights  distance  points accumulated  points redeemed  \
371684              2       406                40.0                0   
371685              0         0                 0.0                0   

        dollar cost points redeemed  
371684                            0  
371685                            0

**En el primer DataFrame, el cliente con el número de fidelidad 992168 tuvo, en el mes de octubre de 2018, dos registros**

El primero, que se puede verificar que es anterior debido a su índice inferior, tuvo un total de 2 vuelos, mientras que el registro con un índice posterior muestra un total de 0 vuelos, lo que podría llevarnos a pensar que sus vuelos fueron cancelados.

In [90]:
duplicated_flight[(duplicated_flight['loyalty number'] == 106001) &(duplicated_flight['month'] == 5) &(duplicated_flight['year'] == 2017)]


loyalty number  year  month  flights booked  flights with companions  \
67719           106001  2017      5              11                        1   
232632          106001  2017      5              14                        7   

        total flights  distance  points accumulated  points redeemed  \
67719              12      3984               398.0                0   
232632             21      3045               304.0              552   

        dollar cost points redeemed  
67719                             0  
232632                           45

**Ya en el segundo DataFrame, en el que el cliente  con el número de fidelidad 106001 viajó en el mes de mayo de 2017**

Podemos observar que la primera línea del DataFrame tiene un total de 12 vuelos, mientras que la segunda línea del DataFrame tiene un total de 21 vuelos. Con la ayuda del índice, podemos notar una diferencia muy grande entre un registro y otro; sin embargo, el registro con el índice más alto podríamos asumir que corresponde al registro más reciente del cliente.

### Posibles causas:

- **Actualización de los datos en momentos diferentes:** Es posible que los datos hayan sido actualizados en distintos momentos, lo que resultó en registros duplicados con información que varió con el tiempo. El índice más alto generalmente indica un registro más reciente, por lo tanto, podría reflejar los datos más actualizados.

- **Error de registro o duplicación:** Puede haber ocurrido un error en el sistema que generó una duplicación, o bien un registro parcial seguido de uno completo.

- **Cancelación o reprogramación:** La diferencia en el número total de vuelos y en los puntos acumulados podría indicar que uno de los registros fue modificado debido a cancelaciones o cambios en los vuelos.

**Verificación de Ocurrencias: 0.49%, no representando una % de errores grandes**

In [91]:
# Step 1: Identify duplicate rows by loyalty number, year and month
duplicate_flight_copy = df_flight[df_flight.duplicated(subset=['loyalty number', 'year', 'month'], keep=False)].copy()
# .copy() is important here to create a true independent copy of the filtered rows

# Step 2: Convert grouping columns to integers using .loc (avoids SettingWithCopyWarning)
duplicate_flight_copy[['loyalty number', 'year', 'month']] = duplicate_flight_copy[['loyalty number', 'year', 'month']].astype(int)

# Step 3: Count unique values of 'total flights' in each group
conflicts = duplicate_flight_copy.groupby(['loyalty number', 'year', 'month'])['total flights'].nunique()

# Step 4: Filter groups with more than one unique 'total flights' value (conflicts)
conflicts = conflicts[conflicts > 1]

# Step 5: Convert result to DataFrame
conflicts_df = conflicts.reset_index(name='conflicting_total_flights')

conflicts_df


loyalty number  year  month  conflicting_total_flights
0             101902  2017      4                          2
1             101902  2017      5                          2
2             101902  2017      6                          2
3             101902  2017      8                          2
4             101902  2017      9                          2
...              ...   ...    ...                        ...
1987          992168  2018      7                          2
1988          992168  2018      8                          2
1989          992168  2018     10                          2
1990          992168  2018     11                          2
1991          992168  2018     12                          2

[1992 rows x 4 columns]

In [92]:
# Total number of rows
total_lines = df_flight.shape[0]

# Number of customers with duplicated records
duplicated_customers = 1992

# Calculating the percentage
percentage_duplicated = round((duplicated_customers / total_lines) * 100, 2)

# Print result with percentage formatted to 2 decimal places
print(f'The percentage of customers with duplications is {percentage_duplicated}%')


The percentage of customers with duplications is 0.49%


### Dado que existen estas diferencias, optaré por mantener el registro más reciente (el que tiene el índice más alto), ya que probablemente representa la información más actual. 

### No obstante, consultaré con la persona responsable del origen de los datos para saber si dispone de información adicional que ayude a confirmar esta conclusión.

In [93]:
df_flight.sort_index(inplace=True)  # Ensures the most recent row (highest index) comes last 
# Use inplace=True when: You want to apply the change directly and don’t need the previous state of the DataFrame
# Use inplace=False when:You want to test a transformation without losing the original DataFrame

df_flight.drop_duplicates(subset=['loyalty number', 'year', 'month'], keep='last', inplace=True)


| Valor de `keep` | Qué hace                                                  |
|------------------|-----------------------------------------------------------|
| `'first'`        | Mantiene la **primera ocurrencia** y elimina las demás    |
| `'last'`         | Mantiene la **última ocurrencia** y elimina las demás ✅  |
| `False`          | Elimina **todas las duplicadas**, sin conservar ninguna   |


In [94]:
# Using keep=False is ideal when you want to detect all duplicated records, not just the repeated ones.
df_flight.duplicated(subset=['loyalty number', 'year', 'month'], keep=False).sum()



np.int64(0)

**Dataframe: Customer Loyalt History**

In [95]:
# Loyalty DataFrame: Zero duplicates
df_loyalty.duplicated().sum()


np.int64(0)

### 7. Descriptive statistics for both numeric and object columns  


In [96]:
# df_flight
round(df_flight.describe().T, 2)

count       mean        std       min  \
loyalty number               401688.0  549735.88  258904.72  100018.0   
year                         401688.0    2017.50       0.50    2017.0   
month                        401688.0       6.50       3.45       1.0   
flights booked               401688.0       4.11       5.22       0.0   
flights with companions      401688.0       1.03       2.08       0.0   
total flights                401688.0       5.14       6.52       0.0   
distance                     401688.0    1208.87    1432.93       0.0   
points accumulated           401688.0     123.70     146.59       0.0   
points redeemed              401688.0      30.75     125.62       0.0   
dollar cost points redeemed  401688.0       2.49      10.16       0.0   

                                   25%       50%        75%       max  
loyalty number               326603.00  550434.0  772019.00  999986.0  
year                           2017.00    2017.5    2018.00    2018.0  
month                             3.75       6.5       9.25      12.0  
flights booked                    0.00       1.0       8.00      21.0  
flights with companions           0.00       0.0       1.00      11.0  
total flights                     0.00       1.0      10.00      32.0  
distance                          0.00     490.0    2336.00    6293.0  
points accumulated                0.00      50.0     239.00     676.5  
points redeemed                   0.00       0.0       0.00     876.0  
dollar cost points redeemed       0.00       0.0       0.00      71.0

In [97]:
# df_loyalty
round(df_loyalty.describe().T, 2)

count       mean        std        min        25%  \
loyalty number    16737.0  549735.88  258912.13  100018.00  326603.00   
salary            16737.0   77923.60   30096.80   15609.00   63991.00   
clv               16737.0    7988.90    6860.98    1898.01    3980.84   
enrollment year   16737.0    2015.25       1.98    2012.00    2014.00   
enrollment month  16737.0       6.67       3.40       1.00       4.00   

                        50%        75%        max  
loyalty number    550434.00  772019.00  999986.00  
salary             73510.00   82940.00  407228.00  
clv                 5780.18    8940.58   83325.38  
enrollment year     2015.00    2017.00    2018.00  
enrollment month       7.00      10.00      12.00

In [98]:
# df_loyalty
df_loyalty.describe(include='O').T

count unique       top   freq
country          16737      1    Canada  16737
province         16737     11   Ontario   5404
city             16737     29   Toronto   3351
postal code      16737     55   V6E 3D9    911
gender           16737      2    Female   8410
education        16737      5  Bachelor  10475
marital status   16737      3   Married   9735
loyalty card     16737      3      Star   7637
enrollment type  16737      2  Standard  15766

### 8. **Detection of constant columns**  


In [99]:
# Columns with only one unique value (constant columns) in df_fligh
constant_cols1 = [col for col in df_flight.columns if df_flight[col].nunique() <= 1]


In [100]:
# Columns with only one unique value (constant columns) in df_loyalty
constant_cols2 = [col for col in df_loyalty.columns if df_loyalty[col].nunique() <= 1]

In [101]:
df_loyalty['country'].unique()

array(['Canada'], dtype=object)

Todos los registros de la columna `country` indicaban que los clientes pertenecen a **Canadá**, sin ninguna variación.  
Por lo tanto, esa columna **no aporta información útil para el análisis**.

Después de verificar que esta información era redundante, decidí **eliminar la columna `country`** para mantener el DataFrame más limpio y eficiente.

Después de realizar el `merge` con el DataFrame de vuelos renombre el CSV de Customer Loyalty Program & Flight Data – Canada.csv


### 9. **Data type inspection**  


- **df_loyalty.dtypes  -> En un principio sus tipos estan de acuerdo con el contenido del dato**

- **df_flight.dtypes -> son todos int o float**

In [102]:
df_loyalty.dtypes

loyalty number        int64
country              object
province             object
city                 object
postal code          object
gender               object
education            object
salary              float64
marital status       object
loyalty card         object
clv                 float64
enrollment type      object
enrollment year       int64
enrollment month      int64
cancelled              bool
dtype: object

- Year: 2017 y 2018
- Month: 12 months represented by numbers 

In [103]:
# Number of unique values per column (cardinality)
df_flight.nunique().sort_values()


year                               2
month                             12
flights with companions           12
flights booked                    22
total flights                     33
dollar cost points redeemed       49
points redeemed                  587
points accumulated              1549
distance                        4746
loyalty number                 16737
dtype: int64

- Country: Canada
- Enrollment Month: 12 months represented by numbers 
- CLV: 

In [104]:
df_loyalty.nunique().sort_values()

country                 1
gender                  2
cancelled               2
enrollment type         2
marital status          3
loyalty card            3
education               5
enrollment year         7
province               11
enrollment month       12
city                   29
postal code            55
salary               5870
clv                  7984
loyalty number      16737
dtype: int64

### 12. **Merge Dataframes**


Tomé la decisión de usar el método **`merge()` con el parámetro `how='left'`**, que representa una **unión izquierda (left join)**, porque:

- El DataFrame `df_flight` tiene más de 400.000 registros, lo que indica que hay múltiples filas por cliente (histórico de vuelos).
- Usamos `how='left'` para conservar **todas las observaciones de vuelos**, incluso si el cliente ya no aparece en `df_loyalty` (por baja del programa, datos faltantes, etc.).
- Esto permite analizar patrones como el churn, la cancelación, o actividad sin información personal.

In [105]:
# Merge df_flight with df_loyalty using a left join on 'loyalty number'
df_merged = df_flight.merge(df_loyalty, on='loyalty number', how='left')

# Result: keeps all rows from df_flight and adds matching info from df_loyalty


In [106]:
df_merged.columns = df_merged.columns.str.strip().str.title() # Formatacion incial recuperada antes de exportar a Data Frame 
df_merged.to_csv('Customer Loyalty Program & Flight Data – Canada.csv', index=False)
